contracts/starknet/space/space.cairo

```
@@ -0,0 +1,214 @@
+    assert_nn(_voting_delay)
+    assert_nn(_voting_period)
+    assert_not_zero(_voting_strategy)
+    assert_not_zero(_authenticator)
+    # TODO: maybe use uint256_signed_nn to check proposal_threshold?
+
+    # Initialize the storage variables
+    voting_delay.write(_voting_delay)
+    voting_period.write(_voting_period)
+    proposal_threshold.write(_proposal_threshold)
+    voting_strategy.write(_voting_strategy)
+    authenticator.write(_authenticator)
+    next_proposal_nonce.write(1)
+
+    return ()
+end
+
+@external
+func vote{syscall_ptr : felt*, pedersen_ptr : HashBuiltin*, range_check_ptr : felt}(
+        voter_address : EthAddress, proposal_id : felt, choice : felt) -> ():
+    alloc_locals
+
+    # Verify that the caller is the authenticator contract.
+    authenticator_only()
+
+    let (proposal) = proposal_registry.read(proposal_id)
+    let (current_timestamp) = get_block_timestamp()
+
+    # Make sure proposal is not closed
+    assert_lt(current_timestamp, proposal.end_timestamp)
+
+    # Make sure proposal has started
+    assert_le(proposal.start_timestamp, current_timestamp)
+
+    # Make sure voter has not already voted
+    let (prev_vote) = vote_registry.read(proposal_id, voter_address)
+    if prev_vote.choice != 0:
+        # Voter has already voted!
+        assert 1 = 0
+    end
+
+    let (local params : felt*) = alloc()
+    let (strategy_contract) = voting_strategy.read()
+
+    let (user_voting_power) = IVotingStrategy.get_voting_power(
+        contract_address=strategy_contract,
+        timestamp=current_timestamp,
+        address=voter_address,
+        params_len=0,
+        params=params)
+
+    # Make sure `choice` is a valid choice
+    assert_le(Choice.FOR, choice)
+    assert_le(choice, Choice.ABSTAIN)
+
+    let (previous_voting_power) = vote_power.read(proposal_id, choice)
+    # TODO: decide what to do with carry
+    let (new_voting_power, carry) = uint256_add(user_voting_power, previous_voting_power)
```

We should throw an assertion error if the carry is non zero. It is then up to the strategy creators to create strategies that yield voting powers that wont violate this.

---

[Throw if overflow happens](https://github.com/snapshot-labs/sx-core/pull/47/commits/7e8bc4562c966e2e46807f2210859a4e565158e7)